File: racket/oo/puzzle/super_dispatch.cc
#include <iostream>

using namespace std;


class c1 {
 public:
  void m1() {
    m2(); // static display for self
  }
  void m2() {
    cout << "c1:m2" << endl;
  }
};

class c2: public c1 {
 public:
  void m1() {
    cout << "c2:m1" << endl;
  }
  void m2() {
    cout << "c2:m2" << endl;
  }
  void m3() {
    c1::m1(); // c++ explicitly specify the super class.
  }
};

class c3: public c2 {
  void m1() {
  }
  void m2() {
  }
};

int main(int argc, const char *argv[])
{
  c3* o3 = new c3();
  o3->m3();
  return 0;
}

File: racket/oo/puzzle/super_dispatch.py
class c1(object):
    def m1(self):
        self.m2() # dynamic dispatch

    def m2(self):
        print "c1:m2"

class c2(c1):
    def m1(self):
        print "c2:m1"

    def m2(self):
        print "c2:m2"

    def m3(self):
        super(c2, self).m1() # static dispatch

class c3(c2):
    def m1(self):
        print "c3:m1"

    def m2(self):
        print "c3:m2"


o3 = c3()
o3.m3()
